Dorska/video analytics frontend

adorsk opened the pull request:
#### What are the relevant tickets?
https://github.com/mitodl/odl-video-service/issues/402
https://github.com/mitodl/odl-video-service/issues/403

#### What's this PR do?
Adds frontend code for video analytics feature.
see: https://github.com/mitodl/odl-video-service/issues/402

#### How should this be manually tested?

##### 1. Test collections page w/ mock data
1. Go to a collection page that lists videos.
2. Open a developer console, and set `window.ovsMockAnalytics = 1`. This will cause the frontend to query the backend for 'mock' analytics data.
3. Click the triple-dot menu on any video card, and select `Analytics`. The analytics dialog should pop-up.
4. Ensure that the chart/table look reasonable, per 
https://github.com/mitodl/odl-video-service/issues/402#issuecomment-374749212

##### 2. Test VideoDetail page w/ mock data
1. Go to a video detail page (the page that has the video player)
2. Open a developer console, and set `window.ovsMockAnalytics = 1`. This will cause the frontend to query the backend for 'mock' analytics data.
3. Click the `Analytics` button. The analytics dialog should pop-up.
4. Ensure that the chart/table look reasonable, per 
https://github.com/mitodl/odl-video-service/issues/402#issuecomment-374749212

##### 3. Test w/ real data.
0. Setup a google analytics 'property'.
1. Add a dimension to your GA property. Call it 'dimesion1'. See: https://github.com/mitodl/odl-video-service/issues/401 . 
2. Setup a google service account and save its keyfile.
3. Add your service account to your GA property. (In GA, 'admin' => 'user management' under the property section).
4. add these settings to your `.env` file:
  - `GA_CAMERA_DIMENSION=dimension1` (from dimension above)
  - `GA_TRACKING_ID` (e.g. from GA account, `UA-115967001-1`)
  - `GA_VIEW_ID` (e.g. `171841111`) (from GA view)
  - `GA_KEYFILE_JSON` (json string for the service account keyfile)
5. Restart your local web server.
6. Go to Google Analytics, and verify that events were logged for your video watching.
7. Bring up an analytics dialog for your video. It should now display the real data, queried from Google Analytics.

#### Where should the reviewer start?
The meat of this feature is in:
`static/js/components/dialogs/AnalyticsDialog`
`static/js/components/analytics`
`static/js/reducers/videoAnalytics`
`static/js/lib/api.js`

mbertrand commented:
@adorsk running `docker-compose run watch npm run-script fmt` should automatically fix most of the lint errors that the test runner complained about.  You can run tests locally with `./script/test/test_suite.sh`.  Also looks like there's a conflict with master that needs to be resolved.

mbertrand commented on static/js/reducers/videoAnalytics_test.js:
```diff
@@ -0,0 +1,44 @@
+
+import rootReducer from "../reducers"
+import { actions } from "../actions"
+import { makeVideo } from "../factories/video"
```
You can remove this import (not used and will get rid of a lint error)

adorsk commented:
@mbertrand Re: textfield width, nice catch. I introduced this regression in #513 .
PR to fix is: https://github.com/mitodl/odl-video-service/pull/528

adorsk commented:
@mbertrand 

### Re: Camera labeling
Added a label generator fn in AnalyticsPane to transform 0-based camera names to 1-based names.

### Re: Clipped chart tooltips
Adjusted chart padding/size so that there's space for tooltips

### Re: linting/formatting
Linted and formatted per the commands you suggested above (handy!)

It's ready for another look now.

mbertrand commented on static/js/components/analytics/AnalyticsPane.js:
```diff
@@ -0,0 +1,177 @@
+
+  generateLabelForChannel(channel: string) {
+    // Hack to make zero-based camera names more suitable for humans.
+    if (!channel) {
```
On closer look, I think the real analytics data is already labeled `camera1` through `camera4`, and it's just the mock data that is 0-3?

adorsk replied:
@mbertrand ah, you're right, the cameras are 1-4:
https://github.com/mitodl/odl-video-service/pull/519/files#diff-3ba30cee865fa843dc5141d396244367L37

I'll change the mock data and take out the extra label formatting.

adorsk commented on static/js/containers/CollectionDetailPage.js:
```diff
@@ -215,12 +238,25 @@ const mapStateToProps = (state, ownProps) => {
-    { name: DIALOGS.COLLECTION_FORM, component: CollectionFormDialog },
-    { name: DIALOGS.EDIT_VIDEO, component: EditVideoFormDialog },
-    { name: DIALOGS.SHARE_VIDEO, component: ShareVideoDialog },
-    { name: DIALOGS.DELETE_VIDEO, component: DeleteVideoDialog }
```
@mbertrand 

The reason for adding (1) `getDialogComponent`, and (2) `ConnectedCollectionDetailPage` is to simplify testing.

Initially I did use your suggestion:
```
{ name: DIALOGS.ANALYTICS, component: AnalyticsDialog }
```

But with that approach, it was difficult to test the `CollectionDetailPage` component in `CollectionDetailPage_test.js`.

The difficulties arose from having to mock methods/api in `CollectionDetailPage_test.js` related related to analytics components.

I thought it would be bad to mock the analytics methods in `CollectionDetailPage_test.js`, so I added the `getDialogComponent`, and unconnected `CollectionDetailPage` component, to enable mocking of dialogs in `CollectionDetailPage_test.js`.

For example:
https://github.com/mitodl/odl-video-service/pull/519/files#diff-1abcece8e6e13d39a500dd59c2fc93eaR197

With this in mind, do you think we should:
A. keep the `ConnectedCollectionDetailPage`/`getDialogComponent` approach
B. revert to the `{ name: DIALOGS.ANALYTICS, component: AnalyticsDialog }` and mock the analytics logic in `CollectionDetailPage_test.js`
C. something else?

mbertrand replied:
In general I'd prefer to avoid adding complexity to application code for unit testing purposes, but I'm fine with keeping option A, assuming that option B would require the unit tests to be rewritten with a different approach.

adorsk replied:
Roger that, I'll go w/ option A & leave it as-is.

adorsk commented:
@mbertrand Nice catch on videos w/ time gaps in their analytics data.
I added data cleanup in `AnalyticsPane`, `AnalyticsChart`, and `AnalyticsTable` to catch these cases.
Ready for review again, thanks for sticking through this beast.

adorsk commented:
@mbertrand fantastic, thanks!
I'll rebase, then squash down & merge it.

adorsk commented:
Closes #402

adorsk commented:
Closes #403
